### ISSUE TITLE
huggingface, Llama 모델 사용 코드 설명서

### ISSUE THREAD

**Issue by CutTheWire** [TreeNut-KR/ChatBot-AI#6]
아래는 `fastapi\src\utils\ai_models\llama_model.py` 파일에 정의된 `LlamaChatModel` 클래스를 AI 경험이 처음인 사람도 이해할 수 있도록 하나하나 설명한 마크다운 문서 예시입니다.

# LlamaChatModel 클래스 설명서

이 문서는 `llama_model.py` 파일에 정의된 `LlamaChatModel` 클래스에 대한 상세한 설명서입니다.  
AI 모델 사용이 처음인 분들도 각 부분의 역할을 이해할 수 있도록 코드의 흐름과 각 메소드의 기능을 자세히 설명합니다.

---

## 목차
1. [파일 개요]
2. [필요 라이브러리 및 모듈]
3. [LlamaChatModel 클래스]
   - [초기화 메소드 (`__init__`)]
   - [토크나이저 로드 (`load_tokenizer`)]
   - [모델 로드 (`load_model`)]
   - [대화 응답 생성 (`generate_response_stream`)]
4. [코드 실행 예제]
5. [추가 참고사항]

---

## 파일 개요
이 파일은 **LlamaChatModel** 클래스를 정의하며, Hugging Face의 Llama 8B 모델을 활용하여 대화 응답을 생성하는 기능을 제공합니다.  
주요 기능은 다음과 같습니다:
- 환경 변수 설정 및 모델 관련 설정 초기화
- 토크나이저와 모델 로드
- 사용자 입력 텍스트를 기반으로 대화 응답을 스트리밍 방식으로 생성

---

## 필요 라이브러리 및 모듈
파일 상단에서 다양한 라이브러리와 모듈을 임포트합니다.  
각 모듈의 역할은 다음과 같습니다:

| 모듈/라이브러리                           | 역할 설명                                                         |
| ----------------------------------------- | ----------------------------------------------------------------- |
| `os`                                      | 파일 및 디렉토리 경로 관리                                         |
| `threading.Thread`                        | 모델 생성 작업을 비동기적으로 실행하기 위한 스레드 처리             |
| `torch`                                   | PyTorch 딥러닝 프레임워크 (모델 연산 및 GPU 사용)                   |
| `transformers`                            | Hugging Face의 모델 및 토크나이저 제공                             |
| `accelerate`                              | 모델의 가속화 및 혼합 정밀도 학습 지원                              |
| `dotenv`                                  | 환경 변수 파일(.env) 로드                                          |
| `typing.Optional`                         | 타입 힌트 지원                                                    |
| `torch.cuda.amp.GradScaler`               | 혼합 정밀도 학습에서의 스케일 조정                                  |
| `transformers.BitsAndBytesConfig`         | 모델 양자화(quantization) 설정                                     |
| `transformers.TextIteratorStreamer`       | 텍스트 생성을 스트리밍 방식으로 받아오기 위한 도구                  |

---

## LlamaChatModel 클래스
이 클래스는 Llama 8B 모델을 사용하여 입력 텍스트로부터 대화 응답을 생성하는 모든 기능을 포함하고 있습니다.

### 초기화 메소드 (`__init__`)
- **환경 변수 로드**: 상위 디렉토리의 `.env` 파일에서 환경 변수를 읽어옵니다.
- **캐시 디렉토리 및 모델 ID 설정**: 모델 파일을 저장할 캐시 디렉토리와 사용할 모델 ID를 지정합니다.
- **디바이스 설정**: GPU(`cuda:0`)를 사용하여 모델을 실행합니다.
- **모델 옵션 설정**: `torch_dtype`, `trust_remote_code`, `device_map`, `quantization_config` 등을 통해 모델 로드 시 필요한 옵션들을 설정합니다.
- **가속기 및 스케일러 초기화**: 혼합 정밀도(fp16) 학습을 위한 Accelerator와 GradScaler를 초기화합니다.
- **토크나이저와 모델 로드**: 각각 `load_tokenizer()`와 `load_model()` 메소드를 호출하여 모델과 토크나이저를 로드합니다.
- **gradient checkpointing**: 메모리 사용량을 줄이기 위해 모델의 gradient checkpointing을 활성화합니다.
- **대화 기록 초기화**: 추후 대화 기록을 저장할 리스트를 초기화합니다.

**코드 예시:**
```python
def __init__(self):
    # 현재 파일의 경로를 기준으로 상위 디렉토리의 .env 파일 로드
    current_dir = os.path.dirname(os.path.abspath(__file__))
    parent_dir = os.path.dirname(current_dir)
    dotenv_path = os.path.join(parent_dir, '.env')
    load_dotenv(dotenv_path)

    # 캐시 디렉토리 및 모델 ID 설정
    self.cache_dir = "./fastapi/ai_model"
    self.model_id = "meta-llama/Llama-3.1-8B-Instruct"
    
    # GPU 사용 (cuda:0)
    self.device = torch.device("cuda:0")
    
    # 모델 로드 옵션 설정 (양자화 등)
    self.model_kwargs = {
        "torch_dtype": torch.float16,
        "trust_remote_code": True,
        "device_map": {"": self.device},
        "quantization_config": BitsAndBytesConfig(
            load_in_4bit=True,
            double_quant=True,  # 추가 양자화 옵션
            compute_dtype=torch.float16
        )
    }
    
    # Hugging Face 인증 토큰 및 가속기, 스케일러 설정
    self.hf_token = os.getenv("HUGGING_FACE_TOKEN")
    self.accelerator = Accelerator(mixed_precision="fp16", device_placement=False)
    self.scaler = GradScaler()
    
    # 토크나이저 및 모델 로드
    print("토크나이저 로드 중...")
    self.tokenizer = self.load_tokenizer()
    print("모델 로드 중...")
    self.model = self.load_model()
    print("모델과 토크나이저 로드 완료!")
    
    # gradient checkpointing 활성화
    self.model.gradient_checkpointing_enable()
    
    # 대화 기록 초기화
    self.conversation_history = []
```

### 토크나이저 로드 (`load_tokenizer`)
이 메소드는 Hugging Face의 `AutoTokenizer`를 사용하여 모델에 맞는 토크나이저를 로드합니다.

- **토크나이저 생성**: 지정된 `model_id`와 인증 토큰을 이용해 토크나이저를 로드합니다.
- **종료 토큰 추가**: 만약 `eos_token_id`가 없다면 `<|endoftext|>` 토큰을 추가합니다.
- **패딩 토큰 설정**: `pad_token_id`를 `eos_token_id`와 동일하게 설정합니다.

**코드 예시:**
```python
def load_tokenizer(self) -> transformers.PreTrainedTokenizerBase:
    tokenizer = transformers.AutoTokenizer.from_pretrained(
        self.model_id,
        token=self.hf_token
    )
    if tokenizer.eos_token_id is None:
        tokenizer.add_special_tokens({'eos_token': '<|endoftext|>'})
    tokenizer.pad_token_id = tokenizer.eos_token_id
    return tokenizer
```

### 모델 로드 (`load_model`)
이 메소드는 `AutoModelForCausalLM`을 사용하여 Llama 모델을 로드합니다.

- **캐시 디렉토리 활용**: 모델 파일은 지정된 `cache_dir`에 저장됩니다.
- **토큰 인증**: Hugging Face의 인증 토큰을 사용합니다.
- **모델 설정 적용**: `model_kwargs`에 정의된 옵션들을 모델 로드시 함께 적용합니다.

**코드 예시:**
```python
def load_model(self) -> transformers.PreTrainedModel:
    model = transformers.AutoModelForCausalLM.from_pretrained(
        self.model_id,
        cache_dir=self.cache_dir,
        token=self.hf_token,
        **self.model_kwargs
    )
    return model
```

### 대화 응답 생성 (`generate_response_stream`)
이 메소드는 입력 텍스트를 받아 모델이 생성하는 응답을 스트리밍 방식으로 반환합니다.

1. **입력 텍스트 토큰화**:  
   - 입력된 텍스트를 모델이 이해할 수 있는 토큰으로 변환합니다.
   - `padding`과 `truncation` 옵션을 통해 입력 길이를 맞춥니다.

2. **Attention Mask 생성**:  
   - 패딩된 부분은 무시하기 위해 `attention_mask`를 생성합니다.

3. **스트리밍 텍스트 생성 준비**:  
   - `TextIteratorStreamer`를 초기화하여 생성된 텍스트를 실시간으로 받아옵니다.

4. **생성 옵션 설정**:  
   - 샘플링(`do_sample`), Top-p, 온도(`temperature`) 등의 옵션을 지정하여 다양한 응답을 생성할 수 있게 합니다.
   - `repetition_penalty`로 반복 생성 방지를 설정합니다.

5. **비동기 생성 실행**:  
   - `Thread`를 사용하여 모델의 텍스트 생성 과정을 비동기적으로 실행합니다.

6. **응답 스트리밍**:  
   - 생성된 텍스트를 실시간으로 `yield`하여 호출자에게 전달합니다.

**코드 예시:**
```python
def generate_response_stream(self, input_text: str):
    # 1. 입력 텍스트를 토큰화하여 tensor로 변환
    input_ids = self.tokenizer.encode(
        text=input_text,
        return_tensors="pt",
        padding=True,
        truncation=True
    ).to(self.device)
    
    # 2. 패딩을 고려한 attention mask 생성
    attention_mask = (input_ids != self.tokenizer.pad_token_id).long().to(self.device)
    
    # 3. 스트리밍 텍스트 생성을 위한 streamer 초기화
    streamer = TextIteratorStreamer(self.tokenizer, skip_prompt=True)
    
    # 4. 텍스트 생성 옵션 설정
    generation_kwargs = {
        "input_ids": input_ids.to(self.device),
        "attention_mask": attention_mask.to(self.device),
        "min_new_tokens": 1,
        "max_new_tokens": 512,
        "do_sample": True,         # 샘플링 활성화
        "top_p": 0.9,              # Top-p 샘플링
        "top_k": 0,                # Top-k 비활성화
        "temperature": 0.7,        # 다양성을 위한 온도 조절
        "eos_token_id": self.tokenizer.eos_token_id,
        "pad_token_id": (
            self.tokenizer.pad_token_id
            if self.tokenizer.pad_token_id is not None
            else self.tokenizer.eos_token_id
        ),
        "repetition_penalty": 1.2,  # 반복 생성 방지 패널티
        "num_return_sequences": 1,  # 한 번에 하나의 시퀀스 생성
        "streamer": streamer        # 스트리밍 활성화
    }
    
    # 5. 비동기적으로 모델의 텍스트 생성 실행 (별도 스레드 사용)
    thread = Thread(target=self.model.generate, kwargs=generation_kwargs)
    thread.start()
    
    # 6. 생성된 텍스트를 스트리밍 방식으로 yield
    for new_text in streamer:
        yield new_text
```

---

## 코드 테스트 실행 예제
아래는 `LlamaChatModel` 클래스를 테스트 형태로 사용하는 간단한 실행 예제입니다.

```python
if __name__ == "__main__":
    # LlamaChatModel 클래스 인스턴스 생성
    chat_model = LlamaChatModel()
    
    # 사용자 입력 예제 (원하는 질문이나 대화문 입력)
    input_text = "안녕하세요, 오늘 날씨는 어떤가요?"
    
    # 생성된 응답을 스트리밍 방식으로 출력
    for response in chat_model.generate_response_stream(input_text):
        print(response, end="")
```

> **참고**: 위 코드를 실행하기 전에 `.env` 파일에 Hugging Face 토큰(`HUGGING_FACE_TOKEN`)을 설정해야 합니다.

---

## 추가 참고사항
- **환경 변수**: 코드 실행 전, 상위 디렉토리에 위치한 `.env` 파일에 Hugging Face 인증 토큰을 설정해야 합니다.
- **GPU 사용**: `self.device`가 `cuda:0`로 설정되어 있으므로, GPU 환경에서 실행하는 것이 좋습니다.
- **모델 다운로드**: 처음 실행 시, Hugging Face에서 모델 파일을 다운로드 받으므로 인터넷 연결이 필요합니다.
- **비동기 처리**: 텍스트 생성은 별도의 스레드에서 처리되므로, 스트리밍 방식으로 결과가 점진적으로 출력됩니다.

---
